### DISCUSSION TITLE
How to define the snoop filter size if using ruby CHI protocol

### DISCUSSION THREAD

**Discussion by FengJiong-123** [gem5/gem5#297] [Q&A]
Hi guys! I just test my simulator using ruby CHI protocol. My compiling and run command are attached as followed:
Compile:
    scons build/ARMCHI/gem5.opt --default=ARM PROTOCOL=CHI SLICC_HTML=True -j32
Run:
    build/ARMCHI/gem5.opt  configs/deprecated/example/se.py -n 16 --l2cache --num-l2caches 16 --num-l3caches 8 --l1i_size 65536 --l1i_assoc 4 --l1d_size 65534 --l1d_assoc 4 --l2_size 2097152 --l2_assoc 8 --l3_size 2097152 --l3_assoc 8 --cmd tests/test-progs/bw_mem_gsr/src/threads-arm --ruby --topology=CustomMesh --chi-config=configs/example/noc_config/2x4.py --network=garnet

As you can see, I define the cache size and association in the command line. **But the question is where and how I can define the directory size?** Because under the CHI protocol, the cache entry and directory entry will be decoupled and allocate/update individually, which means directory size should also be stated in somewhere.

I also search the defination of directory in the .sm code and find that is PerfectCacheMemory which I guess is an infinate cache.
<img width="468" alt="image" src="https://github.com/gem5/gem5/assets/61892174/ba18c6f2-74ec-499c-83fe-be9321590313">
<img width="542" alt="image" src="https://github.com/gem5/gem5/assets/61892174/579b93a2-fb92-437a-b624-718245319e33">

So is directory infinate? If so, I think it may be unreasonable because directory entry eviction will affect the simulation result in fact.

I an really curious about this and appreciate for all of you guys who can answer me!

**Comment by BobbyRBruce** [comment]
@giactra @tiagormk @andysan : Arm People, do you know how to answer this question?

**Comment by JoshRandall** [comment]
Hi all,

Currently there's no back-invalidation support in the gem5 CHI model, so it doesn't support a finite directory (or snoop filter) design. Typically, snoop filters are sized large enough to make evictions rare and maintain performance.

We've implemented back-invalidations and realistic snoop filter designs in our internal branches and we could provide patches in the near future if you are interested in studying these effects.

Thanks,
Josh

**Reply by powerjg** [reply]
I think this would be a *great* contribution to upstream! Maybe we should create an issue to discuss the progress on this. Especially if you would like some feedback on the development process, creating an issue would be the right place to discuss.

**Reply by JoshRandall** [reply]
Hi Jason, issue opened here: https://github.com/gem5/gem5/issues/368

**Reply by gxflying** [reply]
Hi @JoshRandall  ,  
when you said "Currently there's no back-invalidation support in the gem5 CHI model," ,  dose this mean that currently all the chi cache in the cache hirarchy is Non-inclusive , Non-exlusive cache ?  
acctually I can not find a parameter or description in the online doc to address the inclusion policy
maybe it should be state somewhere explicitly ~
thank, and hope your reply

**Reply by tiagormk** [reply]
Hi @gxflying, Joshua means there is no support for back-invalidations generated by the directory/snoop filter, meaning we currently have an "infinite" directory that tracks all cache lines in upstream caches (so no back-invalidations need to be generated). This is independent from the inclusion policies which can be configured for each cache using these params https://github.com/gem5/gem5/blob/stable/src/mem/ruby/protocol/chi/CHI-cache.sm#L133C1-L143C31 they are also documented here https://www.gem5.org/documentation/general_docs/ruby/CHI/#protocol-overview

**Reply by gxflying** [reply]
Hi @tiagormk  
yes, I understood the original meaning of the comment
as we can go far,  
when a cache controller does not support  for "back-invalidations",  it will NOT support strictly inclusive policy eventually
because the strictly inclusive policy depends on the  "back-invalidations",  am I  right ?
the result is that the chi cache model can only support a NINE like policy currently,   is this right too?

**Reply by JoshRandall** [reply]
Hi @gxflying,
I should have been more precise. Currently, it is the directory/snoop filter eviction that is not supported. The invalidation snoop transactions currently support various cache inclusion policies, which are controlled by the parameters Tiago referred to.

**Reply by gxflying** [reply]
it’s clear,  thanks @JoshRandall @tiagormk

**Comment by BobbyRBruce** [comment]
The discussion now has an associated Issue to track the implementation of this feature (#368). There is no longer need for further discussion.
